### PR TITLE
feature/initial

### PR DESCRIPTION
**Pull request by bwalsh**
This PR:
* Parses the [data dictionary spreadsheet](https://github.com/ACED-IDP/ucl-stavrinides/blob/feature/initial/docs/IDP_UCL_VS_data_dictionary.xlsx) into a [python class](https://github.com/ACED-IDP/ucl-stavrinides/blob/feature/initial/ucl_stavrinides/models/submission.py)
* Creates test fixtures from provided [dummy data](https://github.com/ACED-IDP/ucl-stavrinides/tree/feature/initial/tests/fixtures/IDP_UCL_VS_dataset)
* Transforms the test fixtures into [compliant FHIR](https://github.com/ACED-IDP/ucl-stavrinides/tree/feature/initial/tests/fixtures/IDP_UCL_VS_dataset-FHIR)
   * transformation is controlled by [user configurable yaml files](https://github.com/ACED-IDP/ucl-stavrinides/tree/feature/initial/data/resources)
* Creates a [gen3_tracker project](https://github.com/ACED-IDP/ucl-stavrinides/tree/feature/initial/.g3t) for submission
* Adds [tests](https://github.com/ACED-IDP/ucl-stavrinides/tree/feature/initial/tests) for code compliance and transformer results

Paired with:
https://github.com/ACED-IDP/g3t_etl/pull/1

While several [discussion items on the data exist](https://github.com/ACED-IDP/ucl-stavrinides/issues/3),  the provided test data transforms as:

```
$ g3t utilities meta validate tests/fixtures/IDP_UCL_VS_dataset-FHIR/dummy_data_30pid/
resources:
  summary:
    Procedure: 160
    Specimen: 160
    Observation: 6696
    ResearchStudy: 1
    Condition: 30
    ResearchSubject: 30
    Patient: 30
msg: OK

$ g3t utilities meta validate tests/fixtures/IDP_UCL_VS_dataset-FHIR/dummy_data_200pid/
resources:
  summary:
    Procedure: 858
    Specimen: 858
    Observation: 35922
    ResearchStudy: 1
    Condition: 170
    ResearchSubject: 170
    Patient: 170
msg: OK

$ g3t utilities meta validate tests/fixtures/IDP_UCL_VS_dataset-FHIR/dummy_data_500pid/
resources:
  summary:
    Procedure: 1532
    Specimen: 1532
    Observation: 64134
    ResearchStudy: 1
    Condition: 300
    ResearchSubject: 300
    Patient: 300
msg: OK


```

Loading the `tests/fixtures/IDP_UCL_VS_dataset/dummy_data_30pid.csv`  metadata into a test instance produces the following:

image

image

image

### CODE REVIEW

**Review by lbeckman314 — APPROVED** [review]
# Overview

Looks great — examples from the [user guide](https://github.com/ACED-IDP/ucl-stavrinides/blob/feature/initial/user-guide.md) and all tests ran successfully, documentation very clear and well organized!

Question of whether collaborators will have access to SharePoint link in the [docs README](https://github.com/ACED-IDP/ucl-stavrinides/blob/feature/initial/docs/README.md)?

# Running [User Guide](https://github.com/ACED-IDP/ucl-stavrinides/blob/feature/initial/user-guide.md) Examples ✅

## Transforming Data ✅

```sh
➜ ucl_stavrinides transform tests/fixtures/IDP_UCL_VS_dataset/dummy_data_30pid.csv
Transformed tests/fixtures/IDP_UCL_VS_dataset/dummy_data_30pid.csv into META
```

## Validating with g3t ✅

```sh
➜ g3t utilities meta validate
resources:
  summary:
    Procedure: 160
    Specimen: 160
    Observation: 6696
    ResearchStudy: 1
    Condition: 30
    ResearchSubject: 30
    Patient: 30
msg: OK
```

## Creating Python Class with `dictionary` ✅

```sh
➜ ucl_stavrinides dictionary
Transformed docs/IDP_UCL_VS_data_dictionary-IDP_Mapping.xlsx into jsonschema file in data/resources/submission.schema.json
Use this command to generate pydantic model from schema:
datamodel-codegen  --input data/resources/submission.schema.json --input-file-type jsonschema  --output ucl_stavrinides/models/submission.py
```

### Running `datamodel-codegen` ✅

Initially encountered `ImportError`, resolved with this link:
> https://github.com/psf/black/issues/3111#issuecomment-1177362541

```sh
➜ datamodel-codegen  --input data/resources/submission.schema.json --input-file-type jsonschema  --output ucl_stavrinides/models/submission.py
```

### Generated Python class (truncated) ✅

```py
# generated by datamodel-codegen:
#   filename:  submission.schema.json
#   timestamp: 2024-02-10T02:58:19+00:00

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class Submission(BaseModel):
    id: Optional[str] = Field(None, description='Patient ID')
```

# Running Pytest ⚠️

See comment below for fix on minor Flake8 issue (removing unused variable `groups` on [line 68 of simple.py](https://github.com/ACED-IDP/ucl-stavrinides/blob/feature/initial/ucl_stavrinides/transformers/simple.py#L67)).


```sh
➜ pytest tests/ --cov=ucl_stavrinides
==================================== test session starts =====================================
platform darwin -- Python 3.11.6, pytest-8.0.0, pluggy-1.4.0
rootdir: /Users/beckmanl/code/ucl-stavrinides
configfile: pytest.ini
plugins: cov-4.1.0, anyio-4.2.0
collected 5 items

tests/integration/test_extract.py ..                                                   [ 40%]
tests/integration/test_transform.py .                                                  [ 60%]
tests/unit/test_coding_conventions.py .                                                [ 80%]
tests/unit/test_observation_coding_lookup.py .                                         [100%]

---------- coverage: platform darwin, python 3.11.6-final-0 ----------
Name                                              Stmts   Miss  Cover
---------------------------------------------------------------------
ucl_stavrinides/__init__.py                           0      0   100%
ucl_stavrinides/cli.py                               42     42     0%
ucl_stavrinides/models/__init__.py                    0      0   100%
ucl_stavrinides/models/submission.py                 55      0   100%
ucl_stavrinides/models/submission_dictionary.py      34     34     0%
ucl_stavrinides/transformers/__init__.py             79     25    68%
ucl_stavrinides/transformers/simple.py              169     18    89%
---------------------------------------------------------------------
TOTAL                                               379    119    69%


=============================== 5 passed in 157.42s (0:02:37) ================================
```

File: ucl_stavrinides/transformers/simple.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+    # Define a regular expression pattern to match the specified format
+    pattern = r"^(?P<patient_id>[^_]+)_(?P<mri_area>[^_]+)_(?P<time_points>[AB_]+)(?:_(?P<tissue_block>[^_]+))?$"
+
+    # Try to match the pattern with the provided ID
+    match = re.match(pattern, id_str)
+
+    # If there is a match, create a PatientInfo Pydantic model
+    if match:
+        groups = match.groups()
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply ucl_stavrinides/transformers/simple.py]: thanks.   fixed

**Comment by bwalsh** [comment]
> Initially encountered ImportError, resolved with this link:
> https://github.com/psf/black/issues/3111#issuecomment-1177362541
This should help [add click pip spec](https://github.com/ACED-IDP/ucl-stavrinides/pull/2/commits/3418f3563d5c3bc78af5ae81e8b418b821403a80)
@lbeckman314

**Comment by bwalsh** [comment]
[improve tests: coverage at 82%](https://github.com/ACED-IDP/ucl-stavrinides/pull/2/commits/664e2d6db7419526bb437ad28571fb78c640d8f0)
```
---------------------------------------------------------------------
TOTAL                                               379     68    82%

```

**Comment by bwalsh** [comment]
Removed large transient testing fixture files from git
```
tests/fixtures/IDP_UCL_VS_dataset-FHIR/dummy_data_200pid/*.*
tests/fixtures/IDP_UCL_VS_dataset-FHIR/dummy_data_500pid/*.*
```

**Pull request merged**